### ISSUE TITLE
Error on dict.update when using literals vs strings as keys

### ISSUE THREAD

**Issue by armisael** [python/mypy#15789]
**Bug Report**

The following code raises an error, which looks like a bug to me as literals can be treated as strings:

```python
from typing import Literal
a_dict: dict[str, str] = {}
a_dict_with_literals: dict[Literal["foo", "bar"], str] = {"foo": "foobar"}
a_dict.update(a_dict_with_literals)
```

The error thrown is:

```text
main.py:5: error: Argument 1 to "update" of "MutableMapping" has incompatible type "dict[Literal['foo', 'bar'], str]"; expected "SupportsKeysAndGetItem[str, str]"  [arg-type]
```

It's interesting to notice that the following code does not raise any error:

```python
a_dict: dict[str, str] = {}
the_literal: Literal["foo", "bar"] = "foo"
a_dict[the_literal] = "foobar"
```

**To Reproduce**

Links to [mypy Playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=cfe57e2feb6fde701b5d52287483c89e) and [gist](https://gist.github.com/mypy-play/cfe57e2feb6fde701b5d52287483c89e).

- Mypy version used: 1.4.1
- Python version used: 3.11
- Using the default configuration

**Comment by erictraut** [comment]
Mypy is correct in generating an error here. The type parameter `KT` for the `dict` class is invariant, and `dict.update` requires the type to match.

**Comment by ikonst** [comment]
@erictraut Is this a gap in the expressiveness, that you cannot declare `dict.update` to accept that?

**Comment by erictraut** [comment]
I don't know the background of why `dict.update` is declared the way it is in the typeshed stubs. There's a long comment that provides some insights, but there's probably more history. You could search the typeshed change history and issue tracker or post a question in the typeshed discussion forum if you want to dig deeper.
